### DISCUSSION TITLE
ComponentFlow: conditional rendering of result of previous step

### DISCUSSION THREAD

**Discussion by Drevsh** [spring-projects/spring-shell#464] [General]
I have q question regarding conditional rendering of a previous step result and whether something like this is possible. 

I want to display MulitItemSelection/StringInput based a on SingleItemSelection executed in a the same component flow.
The idea I had is that a parse a selection of some services I have (via a "parse" command). After that I have another command ("configure") which shows the user a SingleItemSelection with all the services which have been parsed. The user can select a service which he wants to configure. After the selection multiple questions are asked how the service should be configured. Some of the questions/configuration need to be based on the parsed service definition. For example some services can be backed up while others can't and only for that services a ConfirmationInput should be shown. While this can be done via _.next(ctx -> void)_ I'm also in need to access/retrieve actual data from the service selected in the selection. (i.e. to enabled/edit certain arguments)

Is anything like this supported in the ComponentFlowBuilder? Or would I need to create a sequence of normal component inputs and extract the result item before creating the next step?

I hope the image helps:
![ComponentFlow](https://user-images.githubusercontent.com/7412602/179394898-bbe74dd8-4d48-4d87-82e3-bab8c17ee835.png)

**Comment by jvalkeal** [comment]
Thought I'd try github discussions and move it here. First thanks for trying this part of a shell!

Is your problem trying to make a large flow out of all this or somehow trying to do something in a `configure` command based on something done in a separate `parse` command?

**Comment by jvalkeal** [comment]
ping @Drevsh if you didn't realise this got moved to discussions.

To get you some more ideas, there is one real world example how this flow can be used. There is an experimental project https://github.com/spring-projects-experimental/spring-cli which has a command `spring initializr new` which mimics start.spring.io UI on a CLI side. It currently doesn't use conditional steps but modifies how `dependencies` are shown based on selected `Spring Boot` version. There is a screencast in its [docs](https://spring-projects-experimental.github.io/spring-cli/spring-cli/initializr.html) to see how it works.

It uses `preHandler` to get boot version from a previous step order to show only multi select items compatible with boot version user chose. https://github.com/spring-projects-experimental/spring-cli/blob/fb6fb5b202e88be85f469b2416fd101a6769708e/src/main/java/org/springframework/cli/command/InitializrCommands.java#L263

While result from a component is usually added to a shared context, it's also possible to manually add "stuff" to shared context to get used in later steps. For that there is a `postHandler`.

I'm curious to hear more about your use cases as conditional steps are currently a bit naive as this is a first version of these components and flow system.

Also remember that you don't need flow to use components as those are independent. Flow was mostly created because using a multiple components together needs a quite a bit of low level plumping for things to work together.

**Reply by Drevsh** [reply]
Thanks for the quick reply and yeah this is definitely discussion, thanks for moving it.

The parse command I see as a different command. (I could do it in one but I think that is unnecessary). The configure command I only have available once the parse command has been executed (via the Availability mechanism)

The postHandler seems like it could work but I'd need to try that out and maybe will not work how I intend it since the postHandler method would nee to add new steps to flow (which I don't think is possible?) 

Maybe I can prepare some code via "normal" components to display what I'd like to and add it here some time later today.

**Reply by jvalkeal** [reply]
Looks you're working with interactive shell then jvm is kept alive and you can store stuff somewhere between commands. Implementing branches to flow is probably not that difficult if there's a clear vision how it would look like from builder dsl point of view and how it'd work.

**Reply by Drevsh** [reply]
Here is an example:
https://github.com/Drevsh/spring-shell-demo 

The following code I think I'd like to somehow write with the help of ComponentFlow (but i'm not even sure if that is possible)

```	
        @ShellMethod(key = "configure", value = "Configure service definitions", group = "install")
	public void configure() {
		var service = selectService();
		selectBackupLocation(service);
		configureArguments(service);
	}
```

The `selectBackupLocation` needs the result of the `selectService`, since the question, if a backup should be created depends if the service whould be backed up etc.

The same is the case fore the `configureArguments`. There only the arguments of the service which where selected should be displayed.

**Reply by Drevsh** [reply]
But now writing this it kinda seems hard for me to even specify how a DSL would like like. Especially since you might want the result of different stages to configure/ add new stages conditionally. 

In some stages you might want the result of the very first stage, while in others you might want the result of the second/third/... stage.
